### PR TITLE
[LKP-3] Added test coverage.

### PR DESCRIPTION
**Pull request by lkaminski96**
Commit based on ticket:
https://presence-analyzer-lkaminski.atlassian.net/browse/LKP-3

### CODE REVIEW

File: src/presence_analyzer/tests.py [hunk 1]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
+            utils.interval(
+                sample_start_time,
+                datetime.time(hour=16, minute=25, second=15)
+            ),
+            28515
+        )
+
+    def test_mean(self):
+        """
+        Test calculating mean.
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: Docstring should be more precise, for example: `Test function for calculating mean.`, or `Test mean() function.`

File: src/presence_analyzer/tests.py [hunk 2]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
+        """
+        self.assertEqual(utils.mean([]), 0)
+        self.assertEqual(utils.mean([4.28, -3.20, -1.08]), 0.0)
+        self.assertAlmostEqual(
+            round(utils.mean([0.3, 0.5, 0.2]), 7),
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: Please test the exact result returned by the function:
```
        self.assertAlmostEqual(
            utils.mean([0.3, 0.5, 0.2]),
            0.3333333333333333
        )
```

File: src/presence_analyzer/tests.py [hunk 3]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
+            0.3333333
+        )
+        self.assertAlmostEqual(
+            round(utils.mean([1, 3, 3]), 7),
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: Test case similar to the previous one. There is no need to duplicate tests. If you want to check the function behavior for different types of numbers, just use one case such as:
```
        self.assertEqual(
            utils.mean([-0.3, 1, 0, 2.1]),
            0.7
        )
```

File: src/presence_analyzer/tests.py [hunk 4]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
+            2.3333333
+        )
+        self.assertAlmostEqual(
+            round(utils.mean([-0.3, -0.5, -0.2]), 7),
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: Duplicate test case

File: src/presence_analyzer/tests.py [hunk 5]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
+            0
+        )
+        self.assertEqual(
+            utils.interval(
+                sample_start_time,
+                datetime.time(hour=16, minute=30, second=0)
+            ),
+            28800
+        )
+        self.assertEqual(
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: Looks similar as previous case. redundant.

File: src/presence_analyzer/tests.py [hunk 6]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
+            1200
+        )
+        self.assertEqual(
+            utils.interval(
+                datetime.time(hour=8, minute=30, second=20),
+                datetime.time(hour=8, minute=30, second=50)
+            ),
+            30
+        )
+        self.assertEqual(
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: Looks similar as previous case. redundant.

File: src/presence_analyzer/tests.py [hunk 7]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
+            utils.seconds_since_midnight(
+                datetime.time(hour=1, minute=30, second=30)
+            ),
+            5430
+        )
+
+    def test_interval(self):
+        """
+        Test intervals.
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: Better `Test interval() function.`

File: src/presence_analyzer/tests.py [hunk 8]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
+        """
+        data = utils.get_data()
+        self.assertListEqual(
+            utils.group_by_weekday(data[10]),
+            [[], [30047], [24465], [23705], [], [], []]
+        )
+
+    def test_seconds_since_midnight(self):
+        """
+        Test calculating seconds.
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: Not precise docstring.

File: src/presence_analyzer/tests.py [hunk 9]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
+            ),
+            60
+        )
+        self.assertEqual(
+            utils.seconds_since_midnight(
+                datetime.time(hour=0, minute=0, second=30)
+            ),
+            30
+        )
+        self.assertEqual(
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py:184]: A few duplicate cases above. Consider the case of calculation at the end of the range as:
`datetime.time(hour=23, minute=59, second=59)` (IMO also cover the different cases)

File: src/presence_analyzer/tests.py [hunk 10]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
         self.assertItemsEqual(data[10][sample_date].keys(), ['start', 'end'])
         self.assertEqual(
             data[10][sample_date]['start'],
             datetime.time(9, 39, 5),
         )
+        self.assertTrue(mock_log.debug.called)
+
+    def test_group_by_weekday(self):
+        """
+        Test grouping presence.
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: Docstring is not clear.

File: src/presence_analyzer/tests.py [hunk 11]
```diff
@@ -47,12 +49,66 @@ def test_api_users(self):
...
+        self.assertEqual(resp.status_code, 404)
+        self.assertTrue(mock_log.debug.called)
+
+        resp = self.client.get('/api/v1/presence_weekday/10')
+        self.assertEqual(resp.status_code, 200)
+        data = json.loads(resp.data)
+        self.assertEqual(len(data), 8)
+        self.assertListEqual(data[0], [u'Weekday', u'Presence (s)'])
+        self.assertListEqual(
+            data[1:],
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: You can test the entire list `data` here.

File: src/presence_analyzer/tests.py [hunk 12]
```diff
@@ -47,12 +49,66 @@ def test_api_users(self):
...
+        """
+        resp = self.client.get('/api/v1/presence_weekday/13')
+        self.assertEqual(resp.status_code, 404)
+        self.assertTrue(mock_log.debug.called)
+
+        resp = self.client.get('/api/v1/presence_weekday/10')
+        self.assertEqual(resp.status_code, 200)
+        data = json.loads(resp.data)
+        self.assertEqual(len(data), 8)
+        self.assertListEqual(data[0], [u'Weekday', u'Presence (s)'])
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: You won't need it if you test the entire `date` list at the same time

File: src/presence_analyzer/.coveragerc [hunk 13]
```diff
@@ -0,0 +1,2 @@
+[run]
+source=/home/lukasz/Devel/presence-analyzer-lkaminski/src/presence_analyzer
```
Comment (pwodzinski) [inline src/presence_analyzer/.coveragerc]: We don't want private settings in repo. Remove this file and `.coveragerc` to `.gitignore` file

File: src/presence_analyzer/tests.py [hunk 14]
```diff
@@ -47,12 +49,66 @@ def test_api_users(self):
         Test users listing.
         """
         resp = self.client.get('/api/v1/users')
+
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: remove blank line

File: src/presence_analyzer/tests.py [hunk 15]
```diff
@@ -71,20 +127,127 @@ def tearDown(self):
...
-        self.assertItemsEqual(data.keys(), [10, 11])
+        self.assertItemsEqual(data.keys(), [10, 11, 12])
         sample_date = datetime.date(2013, 9, 10)
         self.assertIn(sample_date, data[10])
         self.assertItemsEqual(data[10][sample_date].keys(), ['start', 'end'])
         self.assertEqual(
             data[10][sample_date]['start'],
             datetime.time(9, 39, 5),
         )
+        self.assertTrue(mock_log.debug.called)
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py:156]: I run coverage and noticed that the test does not cover these lines from utils.py line 59-61:
```
            if len(row) != 4:
                # ignore header and footer lines
                continue
```
It is very simple to add such a condition, it is enough to test less than 4 elements in the data file, example:
```
...
11,2013-09-13,13:16:56,15:04:02
11,2013/09/13,13:16:56,15:04:02
12,2013/09/14,13:16:56
```
I would like you to consider creating a separate file only with damaged data like: `broken_data.csv`. Then define this file in line 18: 
```
TEST_BROKEN_DATA_CSV = os.path.join(
    os.path.dirname(__file__), '..', '..', 'runtime', 'data', 'broken_data.csv'
)
```
And finally you can overwrite settings only for one test using this decorator:
```
@patch.dict('presence_analyzer.main.app.config', {'DATA_CSV': TEST_BROKEN_DATA_CSV})
```
if you create a separate file for damaged data, restore the original values in the file `test_data.csv`

**Review by pwodzinski — CHANGES_REQUESTED** [review]

**Comment by pwodzinski** [comment]
Please add description to this PR

**Comment by lkaminski96** [comment]
Commit based on ticket:
https://presence-analyzer-lkaminski.atlassian.net/browse/LKP-3

File: src/presence_analyzer/tests.py [hunk 16]
```diff
@@ -53,6 +63,67 @@ def test_api_users(self):
         self.assertEqual(len(data), 2)
         self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
 
+    @patch.dict(
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: don't need to overwrite this data, use the `TEST_DATA_CSV` from class setup

File: src/presence_analyzer/tests.py [hunk 17]
```diff
@@ -53,6 +63,67 @@ def test_api_users(self):
...
+                [u'Tue', 30047.0],
+                [u'Wed', 24465.0],
+                [u'Thu', 23705.0],
+                [u'Fri', 0],
+                [u'Sat', 0],
+                [u'Sun', 0]
+            ]
+        )
+
+    @patch.dict(
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: don't need to overwrite this data, use the `TEST_DATA_CSV` from class setup

File: src/presence_analyzer/tests.py [hunk 18]
```diff
@@ -71,20 +142,89 @@ def tearDown(self):
...
+    )
+    @patch('presence_analyzer.utils.log')
+    def test_get_data(self, mock_log):
         """
         Test parsing of CSV file.
         """
         data = utils.get_data()
         self.assertIsInstance(data, dict)
         self.assertItemsEqual(data.keys(), [10, 11])
+        self.assertNotIn([12], data.keys())
```
Comment (pwodzinski) [inline src/presence_analyzer/tests.py]: not needed. all possibilities are already tested line above

**Review by pwodzinski — CHANGES_REQUESTED** [review]

**Review by pwodzinski — APPROVED** [review]

**Pull request merged**